PhysicsTools/Heppy/python/physicsutils/JetReCalibrator.py

```diff
@@ -67,16 +75,27 @@ def getCorrection(self,jet,rho,delta=0,corrector=None):
         corrector.setRho(rho)
         corr = corrector.getCorrection()
         if delta != 0:
-            if not self.JetUncertainty: raise RuntimeError("Jet energy scale uncertainty shifts requested, but not available")
-            self.JetUncertainty.setJetEta(jet.eta())
-            self.JetUncertainty.setJetPt(corr * jet.pt() * jet.rawFactor())
-            try:
-                jet.jetEnergyCorrUncertainty = self.JetUncertainty.getUncertainty(True) 
-            except RuntimeError as r:
-                print "Caught %s when getting uncertainty for jet of pt %.1f, eta %.2f\n" % (r,corr * jet.pt() * jet.rawFactor(),jet.eta())
-                jet.jetEnergyCorrUncertainty = 0.5
-            #print "   jet with corr pt %6.2f has an uncertainty %.2f " % (jet.pt()*jet.rawFactor()*corr, jet.jetEnergyCorrUncertainty)
-            corr *= max(0, 1+delta*jet.jetEnergyCorrUncertainty)
+            if sources:
+                group_shift = 0.
+                for source in sources:
+                    if source not in self.JetUncertaintyBySources:
+                        raise RuntimeError("Jet energy scale uncertainty source requested, but not available")
+                    self.JetUncertaintyBySources[source].setJetEta(jet.eta())
+                    self.JetUncertaintyBySources[source].setJetPt(corr * jet.pt() * jet.rawFactor())
+                    unc = self.JetUncertaintyBySources[source].getUncertainty(True if delta>0. else False)
+                    group_shift += unc*unc
```

Comment:
Does it ever happen that a specific uncertainty is supposed to have anti-correlated effects across different phase space regions?
In that case, squaring them up will break the feature.
If `len(sources) == 1`, when you don't need to do any addition, you could avoid the square and sqrt so that you would preserve any anti-correlation

Reply:
Done